<a id='Top of Page'></a>

# SQL Project

Table of Contents:

<a href='#Step 1 Describe the Goals of the Study'>Step 1 Describe the Goals of the Study</a>

<a href='#Step 2 Study the Tables'>Step 2 Study the Tables</a>

<a href='#Step 3 Make an SQL Query for each of the Tasks and Output the Results in the Notebook'>Step 3 Make an SQL Query for each of the Tasks and Output the Results in the Notebook</a>

<a href='#Step 4 Conclusions and Recommendations'>Step 4 Conclusions and Recommendations</a>



<a id='Step 1 Describe the Goals of the Study'><a/>

## Project Description

The coronavirus took the entire world by surprise, changing everyone's daily routine. City dwellers no longer spent their free time outside, going to cafes and malls; more people were at home, reading books. That attracted the attention of startups that rushed to develop new apps for book lovers. 

I've been given a database of one of the services competing in this market. It contains data on books, publishers, authors, and customer ratings and reviews of books. This information will be used to generate a value proposition for a new product.

In [34]:
# import libraries and establish connection to the database
import psycopg2
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sqlalchemy import create_engine

db_config = {'user': 'praktikum_student',         # user name
             'pwd': 'Sdf4$2;d-d30pp', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the data base

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})


In [2]:
# use this code for SQL queries: pd.io.sql.read_sql(query, con = engine)
# the unction reads a SQL query or database table into a pandas dataframe

<a id='Step 2 Study the Tables'><a/>

## Step 2 Study the Tables

In [35]:
# displaying ERD for the database using markdown

<img src="book_database_ERD.png">

To begin I'm going to get a look at all the tables in the database.

I'll proceed in this order:
1. Books
2. Authors
3. Publishers
4. Ratings
5. Reviews

In [3]:
#querying the books table
books_query = "SELECT * FROM books"

In [4]:
#creating variable for the book table
books_table = pd.io.sql.read_sql(books_query, con = engine)
books_table.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [5]:
#querying the authors table
authors_query = "SELECT * FROM authors"

In [6]:
authors_table = pd.io.sql.read_sql(authors_query, con = engine)
authors_table.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [7]:
#querying the publishers table
publishers_query = "SELECT * FROM publishers"

In [8]:
publishers_table = pd.io.sql.read_sql(publishers_query, con = engine)
publishers_table.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [9]:
#querying the publishers table
ratings_query = "SELECT * FROM ratings"

In [10]:
ratings_table = pd.io.sql.read_sql(ratings_query, con = engine)
ratings_table.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [11]:
#querying the publishers table
reviews_query = "SELECT * FROM reviews"

In [12]:
reviews_table = pd.io.sql.read_sql(reviews_query, con = engine)
reviews_table.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<a id='Step 3 Make an SQL Query for each of the Tasks and Output the Results in the Notebook'><a/>

<a id='Step 3 Make an SQL Query for each of the Tasks and Output the Results in the Notebook'><a/>

## Step 3 Make an SQL Query for each of the Tasks and Output the Results in the Notebook

Tasks:
1. Find the number of books released after January 1, 2000.
2. Find the number of books released after January 1, 2000.
3. Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).
4. Identify the author with the highest average book rating (look only at books with at least 50 ratings).
5. Find the average number of text reviews among users who rated more than 50 books. 

### Task 1: Find the number of books released after January 1, 2000

In [13]:
books_table.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [14]:
#querying the data
books_after_query = "SELECT * FROM books WHERE publication_date > '2000-01-01'"
books_after_2020_01_01 = pd.io.sql.read_sql(books_after_query, con = engine)

In [15]:
print("The number of books released after January 1, 2000 is:", len(books_after_2020_01_01))

The number of books released after January 1, 2000 is: 819


###  Tasks 2: Find the number of user reviews and the average rating for each book

In [16]:
ratings_table.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [17]:
reviews_table.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [18]:
books_table.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [19]:
book_ratings_query = '''
SELECT books.book_id AS book_id, books.title AS title, COUNT(reviews.review_id) AS review_count, AVG(ratings.rating) AS avg_book_rating 
FROM books LEFT JOIN ratings ON ratings.book_id = books.book_id LEFT JOIN reviews ON reviews.book_id = books.book_id 
GROUP BY books.book_id 
ORDER BY review_count DESC;
'''



In [20]:
books_ratings_reviews = pd.io.sql.read_sql(book_ratings_query, con = engine)
display(books_ratings_reviews)

book_id                                              title  review_count  \
0        948                            Twilight (Twilight  #1)          1120   
1        750                The Hobbit  or There and Back Again           528   
2        673                             The Catcher in the Rye           516   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...           492   
4        299  Harry Potter and the Chamber of Secrets (Harry...           480   
..       ...                                                ...           ...   
995      221                          Essential Tales and Poems             0   
996      808                            The Natural Way to Draw             0   
997      191  Disney's Beauty and the Beast (A Little Golden...             0   
998       83    Anne Rice's The Vampire Lestat: A Graphic Novel             0   
999      672   The Cat in the Hat and Other Dr. Seuss Favorites             0   

     avg_book_rating  
0           3.662500  
1           4.125000  
2           3.825581  
3           4.414634  
4           4.287500  
..               ...  
995         4.000000  
996         3.000000  
997         4.000000  
998         3.666667  
999         5.000000  

[1000 rows x 4 columns]

Please see above for the number of reviews per user and the average rating for each book.

### Task 3: Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis)

In [21]:
publishers_table.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [22]:
books_table.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [23]:
publisher_books_query = '''
SELECT publishers.publisher_id AS publisher_id, publishers.publisher AS publisher_name, COUNT(books.book_id) AS book_count 
FROM publishers LEFT JOIN books ON books.publisher_id = publishers.publisher_id 
WHERE books.num_pages > 50 GROUP BY publishers.publisher_id 
ORDER BY book_count DESC LIMIT 1;
'''


In [24]:
publisher_book_count = pd.io.sql.read_sql(publisher_books_query, con = engine)
publisher_book_count.head(10)

publisher_id publisher_name  book_count
0           212  Penguin Books          42

Penguin Books is the publisher with the greatest number of books over 50 pages. 

### Task 4: Identify the author with the highest average book rating (look only at books with at least 50 ratings)

In [25]:
authors_table.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [26]:
books_table.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [27]:
ratings_table.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [28]:
authors_bookratings_query = '''
SELECT books.book_id AS book_id, books.title AS book_title, authors.author_id AS author_id, authors.author AS author_name, AVG(ratings.rating) AS avg_book_rating, COUNT(ratings.rating) AS number_of_ratings 
FROM books LEFT JOIN authors ON authors.author_id = books.author_id LEFT JOIN ratings ON ratings.book_id = books.book_id 
GROUP BY books.book_id, authors.author_id 
HAVING COUNT(ratings.rating_id) >= 50 
ORDER BY avg_book_rating DESC, number_of_ratings DESC LIMIT 5;
'''

In [29]:
author_ratings = pd.io.sql.read_sql(authors_bookratings_query, con = engine)
author_ratings.head(10)

book_id                                         book_title  author_id  \
0      302  Harry Potter and the Prisoner of Azkaban (Harr...        236   
1      722  The Fellowship of the Ring (The Lord of the Ri...        240   
2      299  Harry Potter and the Chamber of Secrets (Harry...        236   
3      656                                     The Book Thief        402   
4      300  Harry Potter and the Half-Blood Prince (Harry ...        236   

                         author_name  avg_book_rating  number_of_ratings  
0         J.K. Rowling/Mary GrandPré         4.414634                 82  
1                     J.R.R. Tolkien         4.391892                 74  
2         J.K. Rowling/Mary GrandPré         4.287500                 80  
3  Markus Zusak/Cao Xuân Việt Khương         4.264151                 53  
4         J.K. Rowling/Mary GrandPré         4.246575                 73

J.K. Rowling has he highest average book rating of 4.414634 for Harry Potter and the Prisoner of Azkaban. This was followed by The Fellowship of the Ring by J.R.R. Tolkien at 4.391892 and then J.K. Rowling again with Harry Potter and the Chamber of Secrets at 4.287500.

### Task 5: Find the average number of text reviews among users who rated more than 50 books

In [30]:
reviews_table.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [31]:
ratings_table.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [32]:
average_text_query = '''
SELECT COUNT(*) / COUNT(distinct username) AS average_reviews 
FROM reviews 
WHERE username IN (SELECT username FROM ratings 
GROUP BY username 
HAVING COUNT(*) > 50)
'''


In [33]:
text_reviews_average = pd.io.sql.read_sql(average_text_query, con = engine)
print(text_reviews_average)

   average_reviews
0               24


The average number of text reviews among users who rated 50 or more books is 24.

<a id='Step 4 Conclusions and Recommendations'><a/>

## Step 4 Conclusions and Recommendations

In conclusion it is clear that the users of this book review service are quite active, which means that this is an ideal market moving forward. I would recommend the company to give exclusive deals to users who review at least 50 books and use that as an incentive to purchase more books and be more active on the site. 

<a href='#Top of Page'>Top of Page<a/>